apps/src/dance/ai/DanceAiClient.js

```diff
@@ -10,93 +10,78 @@ import CachedForegrounds from '@cdo/static/dance/ai/model/cached-spacy-foregroun
  * example: {"backgroundEffect":"sparkles","backgroundColor":"cool","foregroundEffect":"bubbles"}
  */
 export function chooseEffects(emojis) {
-  var foregroundValues = {};
-  var backgroundValues = {};
-  var paletteValues = {};
-  reformatInputs(foregroundValues, CachedForegrounds);
-  reformatInputs(backgroundValues, CachedBackgrounds);
-  reformatInputs(paletteValues, CachedPalettes);
+  // Obtain final summed output weight based off input received
+  const outputTypes = [CachedBackgrounds, CachedForegrounds, CachedPalettes];
+  const outputWeights = [];
+  outputTypes.forEach(outputJson => {
+    outputWeights.push(calculateOutputWeightsVector(emojis, outputJson));
+  });
 
-  const bestBackground = chooseHighestSum(backgroundValues, emojis);
-  const bestForeground = chooseHighestSum(foregroundValues, emojis);
-  const bestPalette = chooseHighestSum(paletteValues, emojis);
+  // Sort and slice top scoring options, mapped to their output identifiers (e.g. [[0.25, 'squiggles'], ...])
+  const numRandomOptions = 3;
+  const outputOptions = [];
+  outputWeights.forEach((weightVector, i) => {
+    outputOptions.push(
+      obtainTopOptions(numRandomOptions, weightVector, outputTypes[i])
+    );
+  });
+
+  // Clarification of which array correlates to which option
+  const backgroundOptions = outputOptions[0];
```

Comment:
From a code organization perspective - might be nice to break out a function like `getEffect(cachedOutput, numOptions)` that returns you the top `numOptions` effects (or randomly chooses from the top effects) based on some input list? Then you could just call `backgroundEffect: getEffect(CachedBackgrounds, numOptions), foregroundEffect: getEffect(CachedForegrounds, numOtions), etc` (can be named whatever, there's probably a better name than `getEffect`)

Reply:
See note above on potential dynamic options provided threshold that is decided post-playtest